### DISCUSSION TITLE
some questions on the article using-litedb-in-an-asp-net-core-api

### DISCUSSION THREAD

**Discussion by mcshaz** [georgekosmidis/blog.georgekosmidis.net#423] [Q&A] [answered]
First and formost Thank you for the great article - very concise and a grat example to get going with.

The questions are:

1.  The use of magic strings on each API endpoint to retieve the relevant document in each method of the service/data access layer - that is ` _liteDb.GetCollection<WeatherForecast>("WeatherForecast")` - wouldn't it be better to serve the typed ILiteCollection from the LiteDbContext like

```
        private readonly LiteDatabase _database;
        public ILiteCollection<WeatherForecast> GetForecast() => _database.GetCollection<WeatherForecast>("WeatherForecast");
```
  or does this have implications when using a singleton instance to serve all web based requests (or other reasons to avoid this pattern)?

2.  What about setting up indexes - in the forecast example one would assume that it may be a requirement to be able to retrive all dates where the temperature was above or below a threshold in degrees C, or all forecasts between certain dates. I am guessing some code like
```
        public LiteDbContext(IOptions<LiteDbOptions> options)
        {
            _database = new LiteDatabase(options.Value.DatabaseLocation);
            var fc = _database.GetCollection<WeatherForecast>("WeatherForecast");
            if (fc.Count() == 0) { seedForecast(fc); }  
        }

        private static void seedForecast(ILiteCollection<WeatherForecast> forecastCollection)
        {
            var forecasts = new[]
            {
                new Forecast
                {
                     Date = new DateOnly(2023, 2, 3)
                     TemperatureC = 20
                }
                new Forecast
                {
                     Date = new DateOnly(2023, 2, 4)
                     TemperatureC = 17
                }
            };
            forecastCollection.InsertBulk( forecasts);
            forecastCollection.EnsureIndex(f => f.Date);
            forecastCollection.EnsureIndex(f => f.TemperatureC);
        }
```
Any thoughts on where, when and how to create the indexes (it seems a very different pattern to SQL where the indexes are created before adding data)?

Thanks again for the article, example repo and thanks for your thoughts

### and 1 samll suggestion
I have been reading the documentation about autoID, and I think some notes would go well in the article and or a comment or 2 in the code.

If the int Id is default(0), then an auto Id is generated (and likewise for Guid types having a Guid.Empty value). When inserting any other int value, the database will insert that Id (or the operation fails if a duplicate).

Importantly and unlike most SQL databases, the next identiy is NOT persisted in the database, so if a record is inserted, assigned an autoId of 3, then deleted, then a completely different record is inserted it may also be given an id of 3 (technically only if the datafile is closed between delete and insert operation, but shutdowns and crashes definately happen). For this reason, it may not be the best choice for a truly restful API. Options then become either using a LiteDb object Id, Guid or implementing some kind of persistance storage yourself and autoincrementing accordingly.

**Comment by georgekosmidis** [comment, marked as answer]
Thank you very much for the comment!

I am sorry for the delay but i am currently on a paternal leave and it will be some time until i return!

I will extend my answer a later time, but here is a quick one - its all from my mobile so i hope i remember the article and sample repo and my comments are correct! 

1. Its best practice for web to use a single instance - you are avoiding expensive file operations (lock, open, close, etc). What you are suggesting seems like returning a new instance of `LiteDatabase` every time, like a transient approach. Is this correct? Please include a bigger sample if i didn't get it! 
2. You are right, my sample needs indexes. I would try to do it once though and not in every request (even the milliseconds of a check is considerable if you mutliply it with thousands of requests). For example you could create the instance for the `AddSingleton` and in there manually create everything needed, if not exists. 

The suggestion is a very nice one and super useful, thank you 😊
